src/ray/core_worker/transport/direct_task_transport.h

```diff
@@ -193,36 +227,100 @@ class CoreWorkerDirectTaskSubmitter {
   // worker using a single lease.
   const uint32_t max_tasks_in_flight_per_worker_;
 
+  // work_stealing_ indicates whether the work stealing mode is enabled
+  const bool work_stealing_;
+
   /// A LeaseEntry struct is used to condense the metadata about a single executor:
   /// (1) The lease client through which the worker should be returned
   /// (2) The expiration time of a worker's lease.
   /// (3) The number of tasks that are currently in flight to the worker
-  /// (4) The resources assigned to the worker
-  /// (5) The SchedulingKey assigned to tasks that will be sent to the worker
+  /// (4) A boolean that indicates whether we have launched a StealTasks request, and we
+  /// are waiting for the stolen tasks (5) The resources assigned to the worker (6) The
+  /// SchedulingKey assigned to tasks that will be sent to the worker
   struct LeaseEntry {
     std::shared_ptr<WorkerLeaseInterface> lease_client;
     int64_t lease_expiration_time;
     uint32_t tasks_in_flight;
+    bool currently_stealing;
+    int64_t stolen_tasks_to_wait_for;
+    uint32_t steal_tasks_request_pending;
```

Comment:
I wonder if we could squash these fields so that we only have to keep track of whether we are currently stealing from a worker (i.e. that the there is a steal request pending). I'm worried about the complexity of trying to keep track of this asynchronous state.

One way to do it would be to return the stolen task IDs in the reply to the steal request. The owner can then look those tasks up using its local hashmap (`executing_tasks_`). The worker can still reply to the PushTaskRequest, but the owner's handler will be a no-op because the task has already been stolen. Would that work?

Reply:
Addressing this in the new PR! See here [#15475](https://github.com/ray-project/ray/pull/15475)